File: algorithms/sort/radix.cpp
#include<bits/stdc++.h>
using namespace std;

void count_sort(vector<int> &arr, int mul){
  int n = arr.size();
  int freq[10] = {0}, a[n];

  for(int i: arr) freq[(i/mul)%10]++;
  
  for(int i = 1; i < 10; i++) freq[i] += freq[i-1];

  for(int i = n-1; i >= 0; i--){
    int k = (arr[i]/mul)%10;
    a[freq[k]-1] = arr[i]; 
    freq[k]--;
  }

  for(int i = 0; i < n; i++) arr[i] = a[i];
}

void radix_sort(vector<int> &arr){
  int maxi = *max_element(arr.begin(), arr.end());
  int mul = 1;
  
  while(maxi){
    count_sort(arr, mul);
    mul *= 10;
    maxi /= 10;
  }
}

int main(){
  vector<int> arr = {10,21,17,34,44,11,654,123};
  radix_sort(arr);
  for (int i: arr) cout << i << " ";
} 

File: algorithms/sort/radix.py
def count_sort(list, mul):
  max_item = max(list)
  count_list = [0] * 10
  output_list = [0] * len(list)

  for i in list:
    index = (i//mul) % 10
    count_list[index] += 1

  for i in range(len(count_list)-1):
    count_list[i + 1] += count_list[i]

  for i in range(len(list) - 1, -1, -1):
    index = (list[i]//mul) % 10
    output_list[count_list[index] - 1] = list[i]
    count_list[index] -= 1

  for i in range(len(list)):
    list[i] = output_list[i]

def radix_sort(list):
  max_item = max(list)
  mul = 1

  while max_item:
    count_sort(list, mul)
    mul *= 10
    max_item /= 10

list = [10,21,17,34,44,11,654,1236]
radix_sort(list)
print(list)
